File: jetbrains/src/main/kotlin/identitybraces/core/TraitSampler.kt
package identitybraces.core

/**
 * Generates representative braces for a preview, without a document to scan.
 *
 * Deciding *what* to draw in a preview is arithmetic on the same weight table the scanner
 * uses, and a preview that disagrees with the scanner would be worse than no preview — it
 * would be confidently wrong about a catalogue nobody can otherwise inspect. So this rolls
 * through [TraitRoll] exactly as [BraceScanner] does. The only thing it invents is the
 * identities, which real braces get from their declaring text.
 */
object TraitSampler {
    /**
     * Fixed, so a sample set does not reshuffle every time a weight changes. With a fresh
     * seed per render, nudging one slider would redraw a completely different set of braces
     * and you could not tell what your change did.
     */
    const val SEED: ULong = 0x9E3779B97F4A7C15uL

    /** The deepest level [depthOf] descends to before coming back out. */
    const val DEEPEST_SAMPLE = 5

    class Sample(val identity: ULong, val depth: Int, val traits: BraceTraits)

    /** Rolls [count] braces against [weights]. */
    fun take(weights: List<TraitWeight>, count: Int): List<Sample> {
        val table = TraitTable(weights)
        return List(maxOf(0, count)) { i ->
            val identity = Hash.mix(SEED, i.toULong())
            Sample(identity, depthOf(i), TraitRoll.roll(identity, table, isUnmatched = false))
        }
    }

    /**
     * A nesting profile that descends and comes back out, so a depth-coloured preview shows
     * a ramp rather than a flat run of one colour.
     */
    fun depthOf(index: Int): Int {
        val i = if (index < 0) -index else index
        val period = DEEPEST_SAMPLE * 2
        val phase = i % period
        return if (phase <= DEEPEST_SAMPLE) phase else period - phase
    }

    /**
     * A brace wearing exactly one trait, whatever that trait's weight is.
     *
     * Built rather than rolled, because the question it answers is "what *is* a cthulhu
     * brace" — asked before deciding whether to enable it. `isDrawn` is forced rather than
     * derived so an unrecognised layer cannot produce a brace the renderer declines to draw.
     */
    fun single(layer: TraitLayer, id: String): BraceTraits = when (layer) {
        TraitLayer.Body -> BraceTraits(body = id, isDrawn = true)
        TraitLayer.Creature -> BraceTraits(creature = id, isDrawn = true)
        TraitLayer.Costume -> BraceTraits(costume = id, isDrawn = true)
        TraitLayer.Motion -> BraceTraits(motion = id, isDrawn = true)
        TraitLayer.Effect -> BraceTraits(effects = listOf(id), isDrawn = true)
    }
}

File: src/IdentityBraces/Core/TraitSampler.cs
using System.Collections.Generic;

namespace IdentityBraces.Core
{
    /// <summary>
    /// Generates representative braces for a preview, without a buffer to scan.
    /// </summary>
    /// <remarks>
    /// <para>
    /// Split out of the options page so it can be tested. The page's job is drawing; deciding
    /// <em>what</em> to draw is arithmetic on the same weight table the scanner uses, and a
    /// preview that disagrees with the scanner would be worse than no preview — it would be
    /// confidently wrong about a catalogue nobody can otherwise inspect.
    /// </para>
    /// <para>
    /// So this rolls through <see cref="TraitRoll"/> exactly as <see cref="BraceScanner"/>
    /// does. The only thing it invents is the identities, which real braces get from their
    /// declaring text.
    /// </para>
    /// </remarks>
    internal static class TraitSampler
    {
        /// <summary>
        /// Fixed, so a sample set does not reshuffle every time a weight changes.
        /// </summary>
        /// <remarks>
        /// This matters more than it sounds. With a fresh seed per render, nudging one slider
        /// would redraw a completely different set of braces and you could not tell what your
        /// change did. Holding the identities still means only the traits that actually depend
        /// on the moved weight change.
        /// </remarks>
        public const ulong Seed = 0x9E3779B97F4A7C15UL;

        /// <summary>The deepest level <see cref="DepthOf"/> descends to before coming back out.</summary>
        public const int DeepestSample = 5;

        public struct Sample
        {
            public ulong Identity;

            /// <summary>A plausible nesting level, for previewing the depth colour mode.</summary>
            public int Depth;

            public BraceTraits Traits;
        }

        /// <summary>Rolls <paramref name="count"/> braces against <paramref name="weights"/>.</summary>
        public static Sample[] Take(IList<TraitWeight> weights, int count)
        {
            if (count < 0)
            {
                count = 0;
            }

            var table = new TraitTable(weights);
            var samples = new Sample[count];

            for (int i = 0; i < count; i++)
            {
                ulong identity = Hash.Mix(Seed, (ulong)i);

                samples[i] = new Sample
                {
                    Identity = identity,
                    Depth = DepthOf(i),
                    Traits = TraitRoll.Roll(identity, table, false),
                };
            }

            return samples;
        }

        /// <summary>
        /// A nesting profile that descends and comes back out, so a depth-coloured preview
        /// shows a ramp rather than a flat run of one colour.
        /// </summary>
        public static int DepthOf(int index)
        {
            if (index < 0)
            {
                index = -index;
            }

            int period = DeepestSample * 2;
            int phase = index % period;
            return phase <= DeepestSample ? phase : period - phase;
        }

        /// <summary>
        /// A brace wearing exactly one trait, whatever that trait's weight is.
        /// </summary>
        /// <remarks>
        /// Built rather than rolled, because the question it answers is "what <em>is</em> a
        /// cthulhu brace" — asked before deciding whether to enable it. Rolling would only show
        /// it once the weight was already turned up, which is the loop the preview exists to
        /// break.
        /// </remarks>
        public static BraceTraits Single(TraitLayer layer, string id)
        {
            var traits = new BraceTraits { Effects = EmptyEffects };

            switch (layer)
            {
                case TraitLayer.Body:
                    traits.Body = id;
                    break;

                case TraitLayer.Creature:
                    traits.Creature = id;
                    break;

                case TraitLayer.Costume:
                    traits.Costume = id;
                    break;

                case TraitLayer.Motion:
                    traits.Motion = id;
                    break;

                default:
                    traits.Effects = new[] { id };
                    break;
            }

            // Forced, not derived. An unrecognised layer would otherwise produce a brace the
            // factory declines to draw, and the preview would silently show an empty cell for
            // a trait that is perfectly real.
            traits.IsDrawn = true;
            return traits;
        }

        private static readonly string[] EmptyEffects = new string[0];
    }
}
